chall09/vsempere.c

```diff
@@ -0,0 +1,160 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h>
+#include <ctype.h>
+
+#define YEAR_UNIT       0
+#define DAY_UNIT        1
+#define HOUR_UNIT       2
+#define MINUTE_UNIT     3
+#define SECOND_UNIT     4
+#define YEAR_SECONDS    31536000
+#define DAY_SECONDS     86400
+#define HOUR_SECONDS    3600
+#define MINUTE_SECONDS  60
+#define MAX_SECONDS     10000000000000000000ul
+
+static void extract_unit(long long *num, int unit_seconds, long *units)
+{
+    *units = *num / unit_seconds;
+    *num = *num - (*units * unit_seconds);
+}
+
+static char *build_particle(long seconds, char *unit, char *use_separator,
+    int *remaining_particles)
+{
+    char    *str;
+
+    str = 0;
+    if (seconds <= 0 || !(str = malloc(sizeof(char) * 25)))
+        return (str);
+
+    if (*use_separator && *remaining_particles > 1)
+        sprintf(str, ", %ld %s", seconds, unit);
+    else if (*use_separator && *remaining_particles == 1)
+        sprintf(str, " and %ld %s", seconds, unit);
+    else
+        sprintf(str, "%ld %s", seconds, unit);
+
+    *use_separator = 1;
+    (*remaining_particles)--;
+
+    return (str);
+}
+
+static char *build_human_time(long unit_values[5], int particles)
```

Comment:
You are using `long` to store your values, but years might exceed `LONG_MAX`. This will cause your program not to work :(